src/compiler/checker.ts

```diff
@@ -26900,6 +26900,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return propType && getConstituentTypeForKeyType(unionType, propType);
     }
 
+    function getMatchingUnionConstituentForArrayLiteral(unionType: UnionType, node: ArrayLiteralExpression) {
+        const resolvedElements = node.elements.map(el => getContextFreeTypeOfExpression(el));
+        for (const type of unionType.types) {
+            if (!isTupleType(type)) continue;

```

Comment:
q: shouldn't this exit the whole process? Or at least when `!isTupleLikeType(type)`?

Reply:
not sure I understand this question.  if the union constituent is not a tuple, it's not valid, but I don't think that means other union constituents couldn't be, so i think we should continue looping?

Reply:
TS is structural and you can create some weird (acceptable) types like `[number, "foo"] | { "1": "bar" }`. This could lead to rejecting some types that should still be considered here. So the way to ignore the problem would be to just (for now?) exit this process.

`isTupleLikeType` point stands, I think. Types like `{ "0": "foo" }` should contribute to autocompletes

Reply:
I was confused when I left my last comment, was thinking this function was something else.  I'm realizing I don't even understand the purpose of `getMatchingUnionConstituentForObjectLiteral` or `getMatchingUnionConstituentForArrayLiteral`.  Do you know what this function should be doing?  I'm just going to remove it from the PR for now, can add something back if it makes sense.